ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -81,7 +89,61 @@ let ntlm_auth uname passwd : (unit, exn) result =
     Ok ()
   with _ -> Error (auth_ex uname)
 
+let get_domain_info_from_db () =
+  Server_helpers.exec_with_new_task "retrieving external auth domain workgroup"
+  @@ fun __context ->
+  let host = Helpers.get_localhost ~__context in
+  let service_name =
+    Db.Host.get_external_auth_service_name ~__context ~self:host
+  in
+  let workgroup, netbios_name =
+    Db.Host.get_external_auth_configuration ~__context ~self:host
+    |> fun config ->
+    (List.assoc_opt "workgroup" config, List.assoc_opt "netbios_name" config)
+  in
+  {service_name; workgroup; netbios_name}
+
 module Ldap = struct
+  module Escape = struct
+    (*
+     * Escape characters according to
+     * https://docs.microsoft.com/en-gb/windows/win32/adsi/search-filter-syntax?redirectedfrom=MSDN#special-characters
+     * *)
+    let reg_star = Str.regexp "*"
+
+    let reg_left_bracket = Str.regexp "("
```

Comment:
Str is not thread safe, and XAPI is a multithreaded application. We should use `re` here, use `Re.Posix` instead. 
In fact I see we've already merged some code that uses Re.Str, which has the same issues as Str, it is not thread safe, because it relies on global internal state (you can see this from the function prototypes too: matched_group doesn't take the regex or the match object as parameter, it just relies on global state).

Reply:
I've opened an issue here to document this upstream https://github.com/ocaml/ocaml-re/issues/194

Reply:
`matched_group` does not takes regex or match object, indicate it use global state some where,
but I am not sure about the using `global_replace` here, it takes regular expression, replacement, and the whole string, 
`val global_replace : regexp -> string -> string -> string`, which looks like does not needs extra state.